### DISCUSSION TITLE
How to access hidden CST tokens from AST?

### DISCUSSION THREAD

**Discussion by enikao** [eclipse-langium/langium#1462] [Q&A]
I'm trying to retrieve the complete input text for a specific AST node, including all hidden tokens.

My example is a properties file:

```
integerProp = 1
myString = "Hello, world!"
```

So if I accessed AST `model.props[0].$cstNode.content`, I'd expect to see 5 tokens: `ID`, `WS`, `EQ`, `WS`, `INT`. But the two `WS` tokens are omitted.
I also tried to poke around in the `model.$document` structure, to no avail.
They seem to exist in `lexerResult.hidden`, but not propagated further?

**Comment by msujew** [comment]
Hey @enikao,

unlike in Xtext, whitespace tokens aren't actually stored in the CST and simply discarded. All other hidden tokens can be accessed via the CST though. You can change that behavior by overriding the `TokenBuilder` and change the group of the white space token to `'hidden'`.

**Reply by cdietrich** [reply]
@msujew when i tried this last time the formatter was completely broken. so i ended up searching gaps between tokens

**Reply by enikao** [reply]
Thanks, I got it working with this adjustment:

```
class PropertiesTokenBuilder extends DefaultTokenBuilder {
    protected override buildTerminalToken(terminal: TerminalRule): TokenType {
        var result = super.buildTerminalToken(terminal);
        if (result.GROUP === Lexer.SKIPPED && isWhitespace(terminalRegex(terminal))) {
            result.GROUP = 'hidden';
        }
        return result;
    }
}
```

@cdietrich I thought about that, but then I can't distinguish between spaces and newlines (or any other kind of whitespace), right?

@msujew I read at [elsewhere](https://github.com/eclipse-langium/langium/discussions/1354#discussioncomment-8231222) you're planning to support full re-serialization. For this to work, do you expect the default setting on whitespace tokens to change?

**Reply by msujew** [reply]
Full reserialization doesn't require whitespace tokens in the CST, as far as I'm aware. I don't expect this to change given the amount of work we had to do in downstream Xtext projects due to incredibly memory hungry CSTs.

**Reply by enikao** [reply]
Given the example

```
prop1=1
prop2     = 2
property3 = 3

prop4=4
    prop5=5
```

How would re-serialization keep the alignment of `prop2`/`property3`, the empty line before `prop4`, and the indentation of `prop5` without storing whitespace?

**Reply by msujew** [reply]
As @cdietrich already mentioned, we use offset information to infer where whitespace was/is supposed to be. If I know that the `prop5` CST node was at `{ line: 5, character: 4 }`, that information can easily be used to infer whitespace information without requiring to store whitespace explicitly. The default formatter implementation already works that way.

**Reply by enikao** [reply]
How does this strategy distinguish tabs vs. spaces for indentation?

**Reply by cdietrich** [reply]
if you do a document.getText(preNonHidden.line/char, nextNonHidden.line/char) it should (not exact api)

**Reply by enikao** [reply]
ah ok, I get it. So it would be a mixture of offset calculation and text slicing.

**Reply by szarnekow** [reply]
>  that information can easily be used to infer whitespace information without requiring to store whitespace explicitly

It's a deliberate decision to ignore windows line endings/unix line endings during full serialization?

**Reply by cdietrich** [reply]
there is no serialization, is there?

**Reply by msujew** [reply]
> there is no serialization, is there?

Right, there is no serialization yet.

**Reply by shutterfreak** [reply]
> Hey @enikao,
> 
> unlike in Xtext, whitespace tokens aren't actually stored in the CST and simply discarded. All other hidden tokens can be accessed via the CST though. You can change that behavior by overriding the `TokenBuilder` and change the group of the white space token to `'hidden'`.

I created a custom TokenBuilder that extends DefaultTokenBuilder with the purpose to give my custom Fomatter access to WS (whitespace) tokens which the DefaultTokenBuilder discards by default. However, when I copy buildTerminalToken() from  DefaultTokenBuilder and use it in my custom TokenBuilder, then the following line does something weird:

```ts
tokenType.GROUP = isWhitespace(regex) ? Lexer.SKIPPED : 'hidden';
```

Instead of setting `tokenType.GROUP` to `'hidden'`, it is set to 'This marks a skipped Token pattern, this means each token identified by it willbe consumed and then thrown into oblivion, this can be used to for example to completely ignore whitespace.'.

All I can come up with, is to explicitly check for is to skip the `RegExpUtils.isWhitespace(regex)` check, essentially removing the entire check. Is that the way to go?

**Reply by msujew** [reply]
> Instead of setting tokenType.GROUP to 'hidden', it is set to 'This marks a skipped Token pattern, this means each token identified by it willbe consumed and then thrown into oblivion, this can be used to for example to completely ignore whitespace.'.

Note that this is the string value inside of `Lexer.SKIPPED`. This sounds like the expected behavior. If you want to change it, go ahead and remove the ternary.
